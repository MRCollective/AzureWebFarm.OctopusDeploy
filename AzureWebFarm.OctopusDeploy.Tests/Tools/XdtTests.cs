using System.IO;
using System.Runtime.CompilerServices;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.Web.XmlTransform;
using Xunit;
using Shouldly;

namespace AzureWebFarm.OctopusDeploy.Tests.Tools
{
    [UseReporter(typeof(DiffReporter))]
    public class XdtTests
    {
        private static readonly string TestPath = Path.GetDirectoryName(typeof (XdtTests).Assembly.CodeBase.Replace("file:///", ""));
        private static readonly string ToolsPath = Path.Combine(TestPath, "Tools");
        private static readonly string ExamplePath = Path.Combine(ToolsPath, "ExampleFiles");

        private static readonly string ExampleCloudProject = Path.Combine(ExamplePath, "CloudProject.ccproj");
        private static readonly string ExampleServiceDefinition = Path.Combine(ExamplePath, "ServiceDefinition.csdef");
        private static readonly string ExampleServiceConfiguration = Path.Combine(ExamplePath, "ServiceConfiguration.cscfg");

        private static readonly string AlreadyTransformedCloudProject = Path.Combine(ExamplePath, "CloudProject.transformed.ccproj");
        private static readonly string AlreadyTransformedServiceDefinition = Path.Combine(ExamplePath, "ServiceDefinition.transformed.csdef");
        private static readonly string AlreadyTransformedServiceConfiguration = Path.Combine(ExamplePath, "ServiceConfiguration.transformed.cscfg");

        private static readonly string XdtCloudProject = Path.Combine(ToolsPath, "CloudProject.ccproj.xdt.xml");
        private static readonly string XdtServiceDefinition = Path.Combine(ToolsPath, "ServiceDefinition.csdef.xdt.xml");
        private static readonly string XdtServiceConfiguration = Path.Combine(ToolsPath, "ServiceConfiguration.cscfg.xdt.xml");

        [Fact]
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void WhenExecutingCloudProjectTransform_ThenCorrectlyTransformExampleFile()
        {
            var result = PerformXdtTransform(ExampleCloudProject, XdtCloudProject);

            Approvals.VerifyXml(result);
        }

        [Fact]
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void WhenExecutingCloudProjectTransform_ThenItShouldBeIdempotent()
        {
            var result = PerformXdtTransform(AlreadyTransformedCloudProject, XdtCloudProject);

            result.ShouldBe(File.ReadAllText(AlreadyTransformedCloudProject));
        }

        [Fact]
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void WhenExecutingServiceDefinitionTransform_ThenCorrectlyTransformExampleFile()
        {
            var result = PerformXdtTransform(ExampleServiceDefinition, XdtServiceDefinition);

            Approvals.VerifyXml(result);
        }

        [Fact]
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void WhenExecutingServiceDefinitionTransform_ThenItShouldBeIdempotent()
        {
            var result = PerformXdtTransform(AlreadyTransformedServiceDefinition, XdtServiceDefinition);

            result.ShouldBe(File.ReadAllText(AlreadyTransformedServiceDefinition));
        }

        [Fact]
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void WhenExecutingServiceConfigurationTransform_ThenCorrectlyTransformExampleFile()
        {
            var result = PerformXdtTransform(ExampleServiceConfiguration, XdtServiceConfiguration);

            Approvals.VerifyXml(result);
        }

        [Fact]
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void WhenExecutingServiceConfigurationTransform_ThenItShouldBeIdempotent()
        {
            var result = PerformXdtTransform(AlreadyTransformedServiceConfiguration, XdtServiceConfiguration);

            result.ShouldBe(File.ReadAllText(AlreadyTransformedServiceConfiguration));
        }

        private static string PerformXdtTransform(string exampleFile, string xdtFile)
        {
            var doc = new XmlTransformableDocument {PreserveWhitespace = true};
            doc.Load(exampleFile);
            var xdt = new XmlTransformation(xdtFile);

            var result = xdt.Apply(doc);

            result.ShouldBe(true);
            using (var stream = new MemoryStream())
            {
                doc.Save(stream);
                stream.Position = 0;
                var reader = new StreamReader(stream);
                return reader.ReadToEnd();
            }
        }
    }
}
