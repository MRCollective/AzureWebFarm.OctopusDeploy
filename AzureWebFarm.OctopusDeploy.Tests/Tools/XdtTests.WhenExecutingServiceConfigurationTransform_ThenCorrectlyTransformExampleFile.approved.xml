<ServiceConfiguration serviceName="WindowsAzure5" xmlns="http://schemas.microsoft.com/ServiceHosting/2008/10/ServiceConfiguration" osFamily="3" osVersion="*" schemaVersion="2013-10.2.2">
  <Role name="WebApplication1">
    <Instances count="1" />
    <ConfigurationSettings>
      <Setting name="Microsoft.WindowsAzure.Plugins.Diagnostics.ConnectionString" value="DefaultEndpointsProtocol=https;AccountName=portalvhdsnxm1kq22c7s2h;AccountKey=9CndQ/phRsHeGGZekc4LhBEilIHD1dMIuaScYrUoVFtMsDo5laj3yoSLPxbua7ebb5L7pBct/rktX/UF+UDc8A==" />
      <Setting name="Microsoft.WindowsAzure.Plugins.RemoteAccess.Enabled" value="true" />
      <Setting name="Microsoft.WindowsAzure.Plugins.RemoteAccess.AccountUsername" value="webfarmadmin" />
      <Setting name="Microsoft.WindowsAzure.Plugins.RemoteAccess.AccountEncryptedPassword" value="MIIBnQYJKoZIhvcNAQcDoIIBjjCCAYoCAQAxggFOMIIBSgIBADAyMB4xHDAaBgNVBAMME1dpbmRvd3MgQXp1cmUgVG9vbHMCEDkO5OmSrouZRmX93GADjMAwDQYJKoZIhvcNAQEBBQAEggEAHFgijEJ2HbG5sc1LX1VJPOso0Jd9e/T2NTXuFzv8tvEqZH3ScmHOh4sC9KHMj01CxCM0tnfL/I4fsj+6AiyPfQld6Z40WGnxQb39o5jl0s74uMOvp6hlLUchRj3DpDPF1XnZ2w4juU7J9a4mq/Gx/Rlk5ASkDOe8twBTAVbi8qmqzg+wFXk7HUq+hl1uaApAS3b079QyE5l/+w6eKgo/jiTV4h7qPorDiSJo76r09/dmgGShMXqsP/nH2FJkhcTeKrD6IfHVslF57sCJZI7nlt/lLEsNSIIp8HJm+wuxY0lYGAgOz7YP4kbDy6sJLl12z92m+s3VBbNk1VkaRa/AuDAzBgkqhkiG9w0BBwEwFAYIKoZIhvcNAwcECEecZEBVk7UBgBDezCCgNAnRbmWldg6QSJt8" />
      <Setting name="Microsoft.WindowsAzure.Plugins.RemoteAccess.AccountExpiration" value="2015-01-02T23:59:59.0000000+08:00" />
      <Setting name="Microsoft.WindowsAzure.Plugins.RemoteForwarder.Enabled" value="true" />
      <Setting name="OctopusServer" value="http://robdmooreoct.cloudapp.net/" />
      <Setting name="OctopusApiKey" value="API-CZYCNXCPAMBWFMDUTPUQUC0O2ZY" />
      <Setting name="TentacleEnvironment" value="Production" />
      <Setting name="TentacleRole" value="webserver" />
      <Setting name="TentacleMachineNameSuffix" value="" />
    </ConfigurationSettings>
    <Certificates>
      <Certificate name="Microsoft.WindowsAzure.Plugins.RemoteAccess.PasswordEncryption" thumbprint="5BFCF449A37F73EF3875A1682BE5AE5CCB1B553F" thumbprintAlgorithm="sha1" />
    </Certificates>
  </Role>
</ServiceConfiguration>